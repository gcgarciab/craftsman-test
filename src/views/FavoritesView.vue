<template>
  <section class="favorites wrapper">
    <h2 class="title">Favorites</h2>

    <div v-if="booksToShow.length > 0" class="favorite-items items-container">
      <BookCard
        v-for="book in booksToShow"
        :key="book.id"
        :id="book.id!"
        :show-detail-link="true"
        :volume-info="book"
        :is-favorite="favorites.includes(book.id!)"
        @toggle-favorite="toggleFavoriteBook(book.id!)"
      />

      <BackToHomeButton />
    </div>

    <template v-else>
      <AppLoader v-if="favorites.length > 0" />

      <ErrorMessage
        v-else
        title="Ups!"
        description="You don't have favorites yet"
        icon="exclamation-triangle"
        :back-to-home="true"
      />
    </template>
  </section>
</template>

<script setup lang="ts">
import { storeToRefs } from 'pinia';
import { computed, onBeforeMount, ref } from 'vue';

import useBookStore from '@/stores/book';
import AppLoader from '@/ui/AppLoader.vue';
import BookCard from '@/components/BookCard.vue';
import ErrorMessage from '@/components/ErrorMessage.vue';
import BackToHomeButton from '@/components/BackToHomeButton.vue';
import BookVolumeInfo from '@/interfaces/book/book-volume-info';

const store = useBookStore();
const { favorites } = storeToRefs(store);
const favoritesData = ref<Partial<BookVolumeInfo>[]>([]);

/**
 * Check if each favorite id is in current favoritesData.
 */
const booksToShow = computed(() =>
  favoritesData.value.filter((book) => favorites.value.includes(book.id!))
);

/**
 * Calls toggleFavoriteBook action from store to add or remove
 * specific book from favorites state.
 * @param id => Book id
 */
function toggleFavoriteBook(id: string) {
  store.toggleFavoriteBook(id);
}

onBeforeMount(async () => {
  favoritesData.value = await store.fetchFavoriteBooks();
});
</script>

<style scoped></style>
