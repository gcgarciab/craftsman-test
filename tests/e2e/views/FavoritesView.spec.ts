import { mount, shallowMount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import { createTestingPinia, TestingOptions } from '@pinia/testing';
import useBookStore from '@/stores/book';
import data from '@/mocks/books-by-id.json';
import FavoritesView from '@/views/FavoritesView.vue';
import findById from '@/utils/find-by-id';

jest.mock('axios', () => ({
  create: jest.fn(() => ({
    get: jest.fn((id: string) => Promise.resolve({ data: findById(data, id) })),
    interceptors: {
      request: { use: jest.fn(), eject: jest.fn() },
      response: { use: jest.fn(), eject: jest.fn() },
    },
  })),
}));

describe('FavoritesView', () => {
  function factory(options?: TestingOptions) {
    const wrapper = mount(FavoritesView, {
      global: {
        plugins: [createTestingPinia(options)],
      },
    });

    const store = useBookStore();
    store.favorites = ['VvJZ3lp2sg0C', 'zt9fCQAAQBAJ', '3yfwAAAAMAAJ'];

    return { wrapper, store };
  }

  it('Call fetchFavoriteBooks store action', async () => {
    const { store, wrapper } = factory({ stubActions: false });

    await flushPromises();
    await flushPromises();
    await flushPromises();
    
    // Await beforemount calls
    expect(store.fetchFavoriteBooks).toBeCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });
});