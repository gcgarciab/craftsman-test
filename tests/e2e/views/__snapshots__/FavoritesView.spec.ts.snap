// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesView Call fetchFavoriteBooks store action 1`] = `
"<section class=\\"favorites wrapper\\">
  <h2 class=\\"title\\">Favorites</h2>
  <div class=\\"h-80 flex justify-center items-center\\">
    <div class=\\"sk-cube-grid\\">
      <div class=\\"sk-cube sk-cube1\\"></div>
      <div class=\\"sk-cube sk-cube2\\"></div>
      <div class=\\"sk-cube sk-cube3\\"></div>
      <div class=\\"sk-cube sk-cube4\\"></div>
      <div class=\\"sk-cube sk-cube5\\"></div>
      <div class=\\"sk-cube sk-cube6\\"></div>
      <div class=\\"sk-cube sk-cube7\\"></div>
      <div class=\\"sk-cube sk-cube8\\"></div>
      <div class=\\"sk-cube sk-cube9\\"></div>
    </div>
  </div>
</section>"
`;
